File: server/src/models/Address.ts
import {
  AfterSave,
  AllowNull,
  BelongsTo,
  Column,
  DataType,
  Default,
  ForeignKey,
  IsUUID,
  Model,
  Not,
  PrimaryKey,
  Table,
} from "sequelize-typescript";
import { ObjectType, Field } from "type-graphql";
import { User } from ".";
import { Transaction } from "sequelize/types";

export enum AddressType {
  BILLING = "billing",
  SHIPPING = "shipping",
}

@ObjectType()
@Table({
  timestamps: true,
  underscored: true,
})
export class Address extends Model {
  @Field()
  @IsUUID(4)
  @PrimaryKey
  @Default(DataType.UUIDV4)
  @Column
  id: string;

  @Field()
  @AllowNull(false)
  @Column
  recipient: string;

  @Field()
  @AllowNull(false)
  @Column
  lineOne: string;

  @Field({ nullable: true })
  @Column
  lineTwo: string;

  @Field()
  @AllowNull(false)
  @Column
  city: string;

  @Field()
  @AllowNull(false)
  @Column
  state: string;

  @Field()
  @AllowNull(false)
  @Column
  zipCode: string;

  @Field()
  @AllowNull(false)
  @Default("United States")
  @Column
  country: string;

  @Field()
  @AllowNull(false)
  @Default(AddressType.SHIPPING)
  @Column
  type: AddressType;

  @AfterSave
  static async ensureOneBilling(
    instance: Address,
    { transaction }: { transaction: Transaction }
  ) {
    if (instance.type === AddressType.BILLING) {
      await Address.update(
        { type: AddressType.SHIPPING },
        {
          where: { userId: instance.userId, id: Not(instance.id) },
          transaction,
        }
      );
    }
  }

  @Field()
  @ForeignKey(() => User)
  @Column
  userId: string;

  @BelongsTo(() => User)
  user: User;

  @Field()
  createdAt: Date;

  @Field()
  updatedAt: Date;
}

File: go-server/internal/models/address.go
package models

import (
	"fmt"

	"gorm.io/gorm"
)

func init() {
	registerModel(&Address{})
}

// AddressType's two values. Note: the Node model never calls type-graphql's
// registerEnumType for this, and a TS *string* enum's emitted reflection
// metadata resolves to GraphQL String — so on the wire `type` has always
// been a plain string, not a real GraphQL enum. The schema matches that.
type AddressType string

const (
	AddressTypeBilling  AddressType = "billing"
	AddressTypeShipping AddressType = "shipping"
)

type Address struct {
	Base

	Recipient string      `gorm:"not null" json:"recipient"`
	LineOne   string      `gorm:"not null" json:"lineOne"`
	LineTwo   string      `json:"lineTwo,omitempty"`
	City      string      `gorm:"not null" json:"city"`
	State     string      `gorm:"not null" json:"state"`
	ZipCode   string      `gorm:"not null" json:"zipCode"`
	Country   string      `gorm:"not null;default:'United States'" json:"country"`
	Type      AddressType `gorm:"not null;default:'shipping'" json:"type"`

	UserID string `gorm:"not null;index" json:"userId"`
	User   *User  `gorm:"foreignKey:UserID;constraint:OnDelete:CASCADE" json:"-"`
}

// AfterSave mirrors Address.ensureOneBilling: at most one address per user
// may be the billing address, so saving one as billing demotes the rest.
func (a *Address) AfterSave(tx *gorm.DB) error {
	if a.Type != AddressTypeBilling {
		return nil
	}
	return tx.Model(&Address{}).
		Where("user_id = ? AND id <> ?", a.UserID, a.ID).
		Update("type", AddressTypeShipping).Error
}

// String matches addressToString in server/src/utils/index.ts.
func (a *Address) String() string {
	str := fmt.Sprintf("%s %s", a.Recipient, a.LineOne)
	if a.LineTwo != "" {
		str += " " + a.LineTwo
	}
	str += fmt.Sprintf(", %s, %s %s, %s", a.City, a.State, a.ZipCode, a.Country)
	return str
}
